### PR TITLE
A number of interop related performance optimizations

### PR DESCRIPTION
**Pull request by sergeystoma**
Hey, 

A project that I'm working on uses Jint to let users write their own business rules to extend functionality. Rules tend to be pretty light on the actual javascript beyond simple math or comparisons with a string manipulation sprinkled around, but instead spend most of the time calling various .NET objects. We've been running into some heavy server load recently, so we took a look if there are any quick wins anywhere, and Jint's interop was the one taking the huge chunk of the time.

I've spent a couple of days optimizing a few things here and there, taking a rather conservative approach without changing significantly the way Jint work. So far I've got interop speed boost of 4-6 times depending on the script. Jint.Benchmark runs about 15% faster, mostly due to conversion of JsValue to ref type. All tests appear to be passing, and we've been using optimized branch in production as well...

Summary of changes:

**[Potentially Breaking?]** JsValue as reference type - removes some excess allocations related to boxing and unboxing, saves a bit of time on reducing copy-by-value penalty
JsValue.FromObject - replaces a cascade of checks with a direct type lookup. Good when you have a lot of different native .NET types which used to always hit worse case scenario of not matching any other type
FieldInfoDescriptor, PropertyInfoDescriptor - **a bit of ugly temporary solution**: fields and property accessors could technically be shared among all instances of the same .NET type. However PropertyDescriptor “owns" the final value, and has to provide it without knowing the object instance that it belongs to. Thus both of these objects used to store Target object reference, thus preventing their reuse. Removed the Target and replaced with ISharedDescriptor workaround… I might revisit this later.

Interop… this is where the biggest chunk of changes is...

DefaultTypeConverter - learns and remembers type->type conversion methods for faster lookups.
**[Potentially Breaking - Portability issues?]** FieldProxy/MethodProxy/PropertyProxy - replaces reflection based accessors with expressions emit-based generated proxies. I think this is quite portable, with exception of iOS /Xamarin (handled as a special case in proxies), but untested
MethodGroup - encapsulates overloaded method resolution
**[Potentially Breaking - see Object.hasOwnProperty note]** TypeInteropDescriptor - works as a pseudo-prototype for .NET types and caches a lot of the information that was otherwise replicated with each ObjectWrapper. Needs to be revisited so it could work as a true prototype. One of the consequences of caching member info separately from ObjectWrapper instances is that Object.hasOwnProperty doesn’t work anymore...

A few optimizations in random places that shaved off a bit of time each or removed chunks of memory allocations (e.g. s1.ToLower()[0] == s2.ToLower()[0] in ObjectWrapper.EqualsIgnoreString)

### CODE REVIEW

**Comment by sebastienros** [comment]
All tests are passing on the CI, that's awesome, and there are no changes to the tests which means the breaking changes are not a huge impact and most apps will still work after the changes, or the changes will be trivial to apply. The perf improvements also show in the CI. That's some great results here, thanks a lot!

I will comment on the code now for you to make some simple changes. Once we agree on the comments then I'll merge.

File: Jint/Engine.cs [hunk 1]
```diff
@@ -42,6 +42,30 @@ public class Engine
 
         // cache of types used when resolving CLR type names
         internal Dictionary<string, Type> TypeCache = new Dictionary<string, Type>();
+		// cache of learned type conversions.
+		internal Dictionary<Type, Func<Engine, object, JsValue>> TypeMappers = new Dictionary<Type, Func<Engine, object, JsValue>>()
```
Comment (sebastienros) [inline Jint/Engine.cs]: What do you think about making it a static concurrent dictionary. That would make the list reusable across engine instances, which is ideal in a server scenario. This applies also to the other caches in this class.

File: Jint/Engine.cs [hunk 2]
```diff
@@ -42,6 +42,30 @@ public class Engine
 
         // cache of types used when resolving CLR type names
         internal Dictionary<string, Type> TypeCache = new Dictionary<string, Type>();
+		// cache of learned type conversions.
```
Comment (sebastienros) [inline Jint/Engine.cs]: Please fix the indentation. There are lots of occurrences like this where you replaced spaces with tabs.

File: Jint/Runtime/Interop/DefaultTypeConverter.cs [hunk 3]
```diff
@@ -11,7 +11,7 @@ namespace Jint.Runtime.Interop
     public class DefaultTypeConverter : ITypeConverter
     {
         private readonly Engine _engine;
-        private static readonly Dictionary<string, bool> _knownConversions = new Dictionary<string, bool>();
+        private static readonly Dictionary<Tuple<Type, Type>, Func<Engine, ITypeConverter, object, IFormatProvider, object>> _knownConversions = new Dictionary<Tuple<Type, Type>, Func<Engine, ITypeConverter, object, IFormatProvider, object>>();
```
Comment (sebastienros) [inline Jint/Runtime/Interop/DefaultTypeConverter.cs]: I don't like Tuples. Please create a custom class/struct.

File: Jint/Runtime/Interop/DefaultTypeConverter.cs [hunk 4]
```diff
@@ -41,182 +41,265 @@ public virtual object Convert(object value, Type type, IFormatProvider formatPro
...
-                        }
-                        var tmpVars = new Expression[@params.Length];
-                        for (var i = 0; i < @params.Count(); i++)
+                        return (Engine engine, ITypeConverter typeConverter, object value, IFormatProvider f) =>
                         {
-                            var param = @params[i];
-                            if (param.Type.IsValueType)
+                            var genericArguments = type.GetGenericArguments();
+
+                            var @params = new ParameterExpression[genericArguments.Count()];
```
Comment (sebastienros) [inline Jint/Runtime/Interop/DefaultTypeConverter.cs]: I don't like variable names with `@`. Please choose another name, like `parameters`, `variables`, ...

Comment (sebastienros) [inline Jint/Runtime/Interop/DefaultTypeConverter.cs]: I realize it's not part of your changes... feel free to leave it if you feel you have done enough.

File: Jint/Engine.cs (hunk 1 above)
Comment (sergeystoma) [inline Jint/Engine.cs]: I'll give this a try, our rules server is already multithreaded, so it would be a good test

**Comment by fatalerrorx** [comment]
JsValue struct -> class is a massive speed boost. Nearly 50% on some tests.

**Comment by sebastienros** [comment]
I don't see any update, if you want I can apply the changes I asked for and still have you as the author of the commit.

**Comment by sergeystoma** [comment]
@sebastienros Hey, sorry. Got overwhelmed with client work lately - I was planning to get back to changes this week, fix the recent merge conflicts and implement proposed changes.

**Comment by sebastienros** [comment]
That's fine, take your time. I might try some of the changes individually too.

**Comment by sebastienros** [comment]
Would it be possible to split this PR into specific ones? Like one for the JsValue as a class, and others for the improvements? I'd like to see the impact in terms of perfs of the JsValue one. It's hard to follow the related changes otherwise.

**Comment by sebastienros** [comment]
Already extracted the `JsValue` changes here: https://github.com/sebastienros/jint/pull/324

**Comment by sebastienros** [comment]
I have merged the changes that include the JsValue refactoring and caching the conversions.
Would be good to see your other changes in another PR though.

**Comment by lahma** [comment]
Closing as relevant bits should be in and on top of that there have been numerous tweaks to improve performance. Please open an new PR if there's something still missing. Thank you!
